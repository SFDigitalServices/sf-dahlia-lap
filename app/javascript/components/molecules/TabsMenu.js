import React from 'react'
import classNames from 'classnames'

import arrayUtils from '~/utils/arrayUtils'
import keyboard from '~/utils/keyboard'

const Tab = ({ title, url, active, onKeyDown, onFocus, onClick, linkRefs }) => {
  const liClassName = classNames({
    'tab-title': true,
    'active': active
  })

  let tabContent
  if (onClick) {
    tabContent =
      <button
        type='button'
        className='button-unstyled'
        onClick={onClick}
        role='menuitem'
        ref={linkRefs}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        tabIndex={active ? '-1' : '0'}
        aria-selected={active}>
        {title}
      </button>
  } else {
    tabContent =
      <a
        href={url}
        className='button-unstyled'
        role='menuitem'
        ref={linkRefs}
        onFocus={onFocus}
        onKeyDown={onKeyDown}
        tabIndex={active ? '-1' : '0'}
        aria-selected={active}>
        {title}
      </a>
  }

  return (
    <li className={liClassName} role='none'>
      {tabContent}
    </li>
  )
}

const TabsMenu = ({ items }) => {
  const tabRefs = arrayUtils.cycle([])

  const handleKeyDown = (e) => {
    keyboard.forEvent(e)
      .on('leftArrow', () => tabRefs.prev().focus())
      .on('rightArrow', () => tabRefs.next().focus())
  }

  const addTabRef = (node) => {
    tabRefs.push(node)
  }

  // NOTE: We need this so arrow navigation and tab navigation play nice together, Fed
  const handleOnFocus = (e) => {
    const idx = tabRefs.values.indexOf(e.target)
    tabRefs.setPosition(idx)
  }

  return (
    <ul className='tabs' role='menubar'>
      {
        items.map((item) => (
          <Tab
            {...item}
            key={item.title}
            onKeyDown={handleKeyDown}
            onFocus={handleOnFocus}
            onClick={item.onClick}
            linkRefs={addTabRef}
            active={item.active} />)
        )
      }
    </ul>
  )
}

export default TabsMenu
