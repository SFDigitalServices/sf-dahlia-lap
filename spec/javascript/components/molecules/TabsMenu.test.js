/* global mount */
import React from 'react'

import TabsMenu from 'components/molecules/TabsMenu'

describe('TabsMenu', () => {
  test('should render tab', () => {
    const items = [
      { title: 'Short Form Application', url: '/url1', active: true },
      { title: 'Supplemental Information', url: '/url2' }
    ]

    const wrapper = mount(
      <TabsMenu items={items} />
    )

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('li').at(0).hasClass('active')).toEqual(true)
    expect(wrapper.find('li').at(1).hasClass('active')).toEqual(false)
  })
})
