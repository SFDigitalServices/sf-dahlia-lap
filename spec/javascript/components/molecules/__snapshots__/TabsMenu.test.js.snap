// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabsMenu should render tab 1`] = `
<TabsMenu
  items={
    Array [
      Object {
        "active": true,
        "title": "Short Form Application",
        "url": "/url1",
      },
      Object {
        "title": "Supplemental Information",
        "url": "/url2",
      },
    ]
  }
>
  <ul
    className="tabs full-width-small-only"
    role="menubar"
  >
    <Tab
      active={true}
      key="Short Form Application"
      linkRefs={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      title="Short Form Application"
      url="/url1"
    >
      <li
        className="tab-title active"
        role="none"
      >
        <a
          aria-selected={true}
          href="/url1"
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1"
        >
          Short Form Application
        </a>
      </li>
    </Tab>
    <Tab
      key="Supplemental Information"
      linkRefs={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      title="Supplemental Information"
      url="/url2"
    >
      <li
        className="tab-title"
        role="none"
      >
        <a
          href="/url2"
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="0"
        >
          Supplemental Information
        </a>
      </li>
    </Tab>
  </ul>
</TabsMenu>
`;
